File: estufainteligente.py
from gpiozero import Button, LED, Servo
from RPLCD.i2c import CharLCD
import time
import analogio  # Supondo uso de placa compatível como Raspberry Pi Pico
import board

# LCD via I2C (ajuste endereço conforme necessário)
lcd = CharLCD('PCF8574', 0x27)

# Servo motor
janela = Servo(board.GP6)

# Sensores analógicos
sensor_temp = analogio.AnalogIn(board.A3)
sensor_solo = analogio.AnalogIn(board.A0)
sensor_ar = analogio.AnalogIn(board.A1)
ldr = analogio.AnalogIn(board.A2)

# Botão com pull-up
botao = Button(board.A4, pull_up=True)

# Atuadores
aquecedor = LED(3)
luz_crescimento = LED(4)
rele = LED(5)
alerta = LED(13)

# Parâmetros ideais
TEMP_IDEAL_MIN = 22.0
TEMP_IDEAL_MAX = 28.0
UMID_AR_IDEAL_MIN = 60.0
UMID_SOLO_SECO = 400
LUZ_IDEAL_MIN = 600

modo_automatico = True
ultimo_monitoramento = time.time()
intervalo_monitoramento = 0.5  # segundos

# Mensagem de modo
tempo_mensagem_modo = 0
duracao_mensagem_modo = 1.0
mostrando_mensagem_modo = False

def analog_to_voltage(pin):
    return pin.value * 3.3 / 65535

def mostrar_mensagem_modo(automatico):
    global mostrando_mensagem_modo, tempo_mensagem_modo
    mostrando_mensagem_modo = True
    tempo_mensagem_modo = time.time()
    lcd.clear()
    lcd.write_string("Modo " + ("Automatico" if automatico else "Manual") + "\nAtivado")

def controlar_temperatura(temp):
    if temp > TEMP_IDEAL_MAX:
        rele.on()
        aquecedor.off()
        janela.max()
    elif temp < TEMP_IDEAL_MIN:
        aquecedor.on()
        rele.off()
        janela.min()
    else:
        aquecedor.off()
        rele.off()
        janela.min()

def controlar_irrigacao(umidade_solo):
    if umidade_solo < UMID_SOLO_SECO:
        print("ACIONANDO BOMBA DE IRRIGACAO")

def controlar_iluminacao(luz):
    if luz < LUZ_IDEAL_MIN:
        luz_crescimento.on()
    else:
        luz_crescimento.off()

def atualizar_display(temp, umid_ar, umid_solo, luz):
    lcd.clear()
    lcd.write_string(f"T:{temp:.1f}C U.A:{int(umid_ar)}%\nU.S:{umid_solo} L:{luz}")

def atualizar_display_inicial():
    lcd.write_string("Aguardando dados\nsensores...")

def verificar_alerta(temp):
    if temp > TEMP_IDEAL_MAX or temp < TEMP_IDEAL_MIN:
        alerta.on()
    else:
        alerta.off()

# Inicialização
lcd.clear()
lcd.write_string("Estufa\nInteligente")
time.sleep(1.5)
lcd.clear()
atualizar_display_inicial()

# Loop principal
while True:
    if botao.is_pressed:
        modo_automatico = not modo_automatico
        mostrar_mensagem_modo(modo_automatico)
        time.sleep(0.2)

    tempo_atual = time.time()

    if mostrando_mensagem_modo and (tempo_atual - tempo_mensagem_modo >= duracao_mensagem_modo):
        mostrando_mensagem_modo = False
        lcd.clear()
        atualizar_display_inicial()

    if not mostrando_mensagem_modo and (tempo_atual - ultimo_monitoramento >= intervalo_monitoramento):
        ultimo_monitoramento = tempo_atual

        tensao = analog_to_voltage(sensor_temp)
        temperatura = (tensao - 0.5) * 100.0

        umidade_solo = sensor_solo.value
        umidade_ar = sensor_ar.value
        luz = ldr.value
        umidade_ar_percentual = int((umidade_ar / 65535) * 100)

        if modo_automatico:
            controlar_temperatura(temperatura)
            controlar_irrigacao(umidade_solo)
            controlar_iluminacao(luz)

        atualizar_display(temperatura, umidade_ar_percentual, umidade_solo, luz)
        verificar_alerta(temperatura)

    time.sleep(0.05)

File: estufainteligente.cpp
#include <LiquidCrystal.h> 

#include <Servo.h> 

  

// LCD: RS, E, D4, D5, D6, D7 

LiquidCrystal lcd(7, 8, 9, 10, 11, 12); 

Servo janela; 

  

// Pinos dos sensores 

const int sensorTempPin = A3;       // TMP36 

const int sensorSoloPin = A0;       // Umidade do solo 

const int sensorArPin = A1;         // Umidade do ar (simulado) 

const int ldrPin = A2;              // Fotoresistor 

const int botaoPin = A4;            // Botão modo automático/manual 

  

// Pinos dos atuadores 

const int aquecedorPin = 3; 

const int luzCrescimentoPin = 4; 

const int relePin = 5; 

const int servoPin = 6; 

const int alertaPin = 13; 

  

// Parâmetros ideais 

const float TEMP_IDEAL_MIN = 22.0; 

const float TEMP_IDEAL_MAX = 28.0; 

const float UMID_AR_IDEAL_MIN = 60.0; 

const int UMID_SOLO_SECO = 400; 

const int LUZ_IDEAL_MIN = 600; 

  

bool modoAutomatico = true; 

unsigned long ultimoMonitoramento = 0; 

const unsigned long intervaloMonitoramento = 500; // 500 ms para atualização rápida 

  

// Variáveis para debounce do botão 

int botaoEstadoAtual = HIGH; 

int botaoEstadoAnterior = HIGH; 

unsigned long ultimoDebounceTime = 0; 

const unsigned long debounceDelay = 10; // 10 ms debounce 

  

// Controle de mensagem de modo no LCD 

unsigned long tempoMensagemModo = 0; 

const unsigned long duracaoMensagemModo = 1000; // 1 segundo mensagem modo 

bool mostrandoMensagemModo = false; 

  

void setup() { 

  lcd.begin(16, 2); 

  janela.attach(servoPin); 

  

  pinMode(aquecedorPin, OUTPUT); 

  pinMode(luzCrescimentoPin, OUTPUT); 

  pinMode(relePin, OUTPUT); 

  pinMode(alertaPin, OUTPUT); 

  pinMode(botaoPin, INPUT_PULLUP); 

  

  Serial.begin(9600); 

  

  // Mensagem inicial no LCD dividida em duas linhas para não cortar 

  lcd.clear(); 

  lcd.setCursor(0, 0); 

  lcd.print("Estufa"); 

  lcd.setCursor(0, 1); 

  lcd.print("Inteligente"); 

  delay(1500); 

  lcd.clear(); 

  

  atualizarDisplayInicial(); 

} 

  

void loop() { 

  // Leitura e debounce do botão 

  int leitura = digitalRead(botaoPin); 

  if (leitura != botaoEstadoAnterior) { 

    ultimoDebounceTime = millis(); 

  } 

  if ((millis() - ultimoDebounceTime) > debounceDelay) { 

    if (leitura != botaoEstadoAtual) { 

      botaoEstadoAtual = leitura; 

      if (botaoEstadoAtual == LOW) { // Botão pressionado 

        modoAutomatico = !modoAutomatico; 

        mostrarMensagemModo(modoAutomatico); 

      } 

    } 

  } 

  botaoEstadoAnterior = leitura; 

  

  unsigned long tempoAtual = millis(); 

  

  // Atualizar sensores e atuadores se não estiver mostrando mensagem modo 

  if (!mostrandoMensagemModo && (tempoAtual - ultimoMonitoramento >= intervaloMonitoramento)) { 

    ultimoMonitoramento = tempoAtual; 

  

    int leituraAnalogica = analogRead(sensorTempPin); 

    float tensao = leituraAnalogica * (5.0 / 1023.0); 

    float temperatura = (tensao - 0.5) * 100.0; 

  

    int umidadeSolo = analogRead(sensorSoloPin); 

    int umidadeAr = analogRead(sensorArPin); 

    int luz = analogRead(ldrPin); 

    float umidadeArPercentual = map(umidadeAr, 0, 1023, 0, 100); 

  

    if (modoAutomatico) { 

      controlarTemperatura(temperatura); 

      controlarIrrigacao(umidadeSolo); 

      controlarIluminacao(luz); 

    } else { 

      // No modo manual, você pode adicionar controle via Serial se quiser 

    } 

  

    atualizarDisplay(temperatura, umidadeArPercentual, umidadeSolo, luz); 

    verificarAlerta(temperatura); 

  } 

  

  // Controla o tempo que a mensagem do modo fica no LCD 

  if (mostrandoMensagemModo && (tempoAtual - tempoMensagemModo >= duracaoMensagemModo)) { 

    mostrandoMensagemModo = false; 

    lcd.clear(); 

    atualizarDisplayInicial(); 

  } 

} 

  

void mostrarMensagemModo(bool automatico) { 

  mostrandoMensagemModo = true; 

  tempoMensagemModo = millis(); 

  lcd.clear(); 

  lcd.setCursor(0, 0); 

  if (automatico) { 

    lcd.print("Modo Automatico"); 

  } else { 

    lcd.print("Modo Manual"); 

  } 

  lcd.setCursor(0, 1); 

  lcd.print("Ativado"); 

} 

  

void controlarTemperatura(float temperatura) { 

  if (temperatura > TEMP_IDEAL_MAX) { 

    digitalWrite(relePin, HIGH); 

    digitalWrite(aquecedorPin, LOW); 

    janela.write(90); 

  } else if (temperatura < TEMP_IDEAL_MIN) { 

    digitalWrite(aquecedorPin, HIGH); 

    digitalWrite(relePin, LOW); 

    janela.write(0); 

  } else { 

    digitalWrite(aquecedorPin, LOW); 

    digitalWrite(relePin, LOW); 

    janela.write(0); 

  } 

} 

  

void controlarIrrigacao(int umidadeSolo) { 

  if (umidadeSolo < UMID_SOLO_SECO) { 

    Serial.println("ACIONANDO BOMBA DE IRRIGACAO"); 

    // Controle da bomba sem delay pode ser implementado aqui 

  } 

} 

  

void controlarIluminacao(int luz) { 

  if (luz < LUZ_IDEAL_MIN) { 

    digitalWrite(luzCrescimentoPin, HIGH); 

  } else { 

    digitalWrite(luzCrescimentoPin, LOW); 

  } 

} 

  

void atualizarDisplay(float temperatura, float umidadeAr, int umidadeSolo, int luz) { 

  lcd.clear(); 

  lcd.setCursor(0, 0); 

  lcd.print("T:"); 

  lcd.print(temperatura, 1); 

  lcd.print("C U.A:"); 

  lcd.print(umidadeAr, 0); 

  lcd.print("%"); 

  

  lcd.setCursor(0, 1); 

  lcd.print("U.S:"); 

  lcd.print(umidadeSolo); 

  lcd.print(" L:"); 

  lcd.print(luz); 

} 

  

void atualizarDisplayInicial() { 

  lcd.setCursor(0, 0); 

  lcd.print("Aguardando dados"); 

  lcd.setCursor(0, 1); 

  lcd.print("sensores..."); 

} 

  

void verificarAlerta(float temperatura) { 

  if (temperatura > TEMP_IDEAL_MAX || temperatura < TEMP_IDEAL_MIN) { 

    digitalWrite(alertaPin, HIGH); 

  } else { 

    digitalWrite(alertaPin, LOW); 

  } 

} 